File: common/python/example_save_device_settings_simple.py
# Copyright 2016 Zurich Instruments AG

"""
Zurich Instruments LabOne Python API Example

Demonstrate how to save and load Zurich Instruments device settings using the
zhinst.utils utility functions save_settings() and load_settings().

Connect to a Zurich Instruments instrument, save the instrument's settings to
file, toggle the signal output enable and reload the settings file.

This example demonstrates the use of the zhinst utility functions
save_settings() and load_settings(). These functions will block until
saving/loading has finished (i.e., they are synchronous functions). Since
this is the desired behaviour in most cases, these utility functions are the
recommended way to save and load settings.

If an asynchronous interface for saving and loading settings is required,
please refer to the `example_save_device_settings_expert'. Which
demonstrates how to directly use the ziDeviceSettings module.

Requirements:
    * LabOne Version >= 20.02
    * Instruments:
        1 x Instrument with demodulators
    * signal output 1 connected to signal input 1 with a BNC cable.

Usage:
    example_save_device_settings_simple.py [options] <device_id>
    example_save_device_settings_simple.py -h | --help

Arguments:
    <device_id>  The ID of the device [device_type: .*LI|.*IA|.*IS]

Options:
    -h --help          Show this screen.
    -s --setting PATH  Specify the path where to save the settings file. [default: ]

Returns:
    filename (str)  the name (with path) of the XML file where the settings were saved.

Raises:
    Exception     If the specified devices do not match the requirements.
    RuntimeError  If the devices is not "discoverable" from the API.

See the "LabOne Programming Manual" for further help, available:
    - On Windows via the Start-Menu:
    Programs -> Zurich Instruments -> Documentation
    - On Linux in the LabOne .tar.gz archive in the "Documentation"
    sub-folder.
"""

import time
import os
import zhinst.utils


def run_example(device_id: str, setting: str = ""):
    """run the example."""

    apilevel_example = 6  # The API level supported by this example.
    # Call a zhinst utility function that returns:
    # - an API session `daq` in order to communicate with devices via the data server.
    # - the device ID string that specifies the device branch in the server's node hierarchy.
    # - the device's discovery properties.
    (daq, device, _) = zhinst.utils.create_api_session(device_id, apilevel_example)
    zhinst.utils.api_server_version_check(daq)

    timestr = time.strftime("%Y%m%d_%H%M%S")
    filename = (
        timestr + "_example_save_device_settings_simple.xml"
    )  # Change this to the filename you want to save.
    if setting:
        print("setting:")
        print(setting)
        filename = setting + os.sep + filename

    toggle_device_setting(daq, device)

    # Save the instrument's current settings.
    print("Saving settings...")
    zhinst.utils.save_settings(daq, device, filename)
    print("Done.")

    # Check we actually saved the file
    assert os.path.isfile(filename), "Failed to save settings file '%s'" % filename
    print(f"Saved file '{filename}'.")

    toggle_device_setting(daq, device)

    # Load settings.
    print("Loading settings...")
    zhinst.utils.load_settings(daq, device, filename)
    print("Done.")

    return filename


def toggle_device_setting(daq, device):
    """
    Toggle a setting on the device: If it's enabled, disable the setting, and
    vice versa.
    """
    path = "/%s/sigouts/0/on" % device
    is_enabled = daq.getInt(path)
    print(f"Toggling setting '{path}'.")
    daq.setInt(path, not is_enabled)
    daq.sync()


if __name__ == "__main__":
    import sys
    from pathlib import Path

    cli_util_path = Path(__file__).resolve().parent / "../../utils/python"
    sys.path.insert(0, str(cli_util_path))
    cli_utils = __import__("cli_utils")
    cli_utils.run_commandline(run_example, __doc__)
    sys.path.remove(str(cli_util_path))

File: hf2-mf-uhf/matlab/example_save_device_settings_simple.m
function filename_fullpath = example_save_device_settings_simple(device_id, varargin)
% EXAMPLE_SAVE_DEVICE_SETTINGS_SIMPLE Save and load device settings to and from file using utility functions
%
% USAGE example_save_device_settings_simple(DEVICE_ID)
%
% Connect to a Zurich Instruments instrument, save the instrument's settings
% to file, toggle the signal output enable and reload the settings
% file. Specify the device to run the example on via by DEVICE_ID, e.g.,
% 'dev1000' or 'uhf-dev1000'.
%
% This example demonstrates the use of the Utility functions ziSaveSettings()
% and ziLoadSettings(). These functions will block until saving/loading has
% finished (i.e., they are synchronous functions). Since this is the desired
% behaviour in most cases, these utility functions are the recommended way to
% save and load settings.
%
% If an asynchronous interface for saving and loading settings is required,
% please refer to the `example_save_device_settings_expert'. Which
% demonstrates how to directly use the ziDeviceSettings module.
%
% NOTE Please ensure that the ziDAQ folders 'Driver' and 'Utils' are in your
% Matlab path. To do this (temporarily) for one Matlab session please navigate
% to the ziDAQ base folder containing the 'Driver', 'Examples' and 'Utils'
% subfolders and run the Matlab function ziAddPath().
% >>> ziAddPath;
%
% Use either of the commands:
% >>> help ziDAQ
% >>> doc ziDAQ
% in the Matlab command window to obtain help on all available ziDAQ commands.
%
% Copyright 2008-2018 Zurich Instruments AG

clear ziDAQ;

if ~exist('device_id', 'var')
    error(['No value for device_id specified. The first argument to the ' ...
           'example should be the device ID on which to run the example, ' ...
           'e.g. ''dev1000'' or ''uhf-dev1000''.'])
end

% Check the ziDAQ MEX (DLL) and Utility functions can be found in Matlab's path.
if ~(exist('ziDAQ') == 3) && ~(exist('ziCreateAPISession', 'file') == 2)
    fprintf('Failed to either find the ziDAQ mex file or ziDevices() utility.\n')
    fprintf('Please configure your path using the ziDAQ function ziAddPath().\n')
    fprintf('This can be found in the API subfolder of your LabOne installation.\n');
    fprintf('On Windows this is typically:\n');
    fprintf('C:\\Program Files\\Zurich Instruments\\LabOne\\API\\MATLAB2012\\\n');
    return
end

% The API level supported by this example.
apilevel_example = 6;
% Create an API session; connect to the correct Data Server for the device.
[device, props] = ziCreateAPISession(device_id, apilevel_example);
ziApiServerVersionCheck();

% Define parameters relevant to this example. Default values specified by the
% inputParser below are overwritten if specified as name-value pairs via the
% `varargin` input argument.
p = inputParser;
% The xml file (without extension) to save instrument settings to.
default_filename_fullpath = [datestr(now, 'yyyymmdd_HHMM') '_settings.xml'];
p.addParamValue('filename_fullpath', default_filename_fullpath, @isstr);
p.parse(varargin{:});

toggleDeviceSetting(device);

fprintf('Saving settings...\n');
ziSaveSettings(device, p.Results.filename_fullpath);
fprintf('Saved file: ''%s''.\n', p.Results.filename_fullpath);

toggleDeviceSetting(device);

fprintf('Loading settings...\n');
ziLoadSettings(device, p.Results.filename_fullpath);
fprintf('Done.\n');

filename_fullpath = p.Results.filename_fullpath;

end


function toggleDeviceSetting(device)
path = ['/' device '/sigouts/0/on'];
on = ziDAQ('getInt', path);
if (on)
    on = 0;
else
    on = 1;
end
fprintf('Toggling ''%s''.\n', path);
ziDAQ('setInt', path, on);
ziDAQ('sync');
end
